### PR TITLE
ui: properly segregate/aggregate app sql stats

### PR DESCRIPTION
**Pull request by couchand**
Previously we had been a bit lazy about the apps that statement statistics roll into, but feedback highlighted that the app was an important distinction for users.  This adds support for apps to the statements list and details pages.

The statements list now has a filter above the list to choose an app:
screen shot 2018-06-25 at 1 04 56 pm

The details page now lists all apps a query appears in (when viewing all apps) or just the stats relevant to a single app.
screen shot 2018-06-25 at 1 05 05 pm

Fixes: #26990

### CODE REVIEW

**Comment by couchand** [comment]
The code and git history are admittedly a bit messy, I want to keep this moving along so I've opened the PR in the current form, happy to take any and all feedback @vilterp.

@knz can you take a look at this and see if it's getting us closer to what you'd expect here?  Thanks!

**Comment by vilterp** [comment]
Saw a runtime error while running TPCC:

http://localhost:8080/#/statement/(unset)/SELECT%20s_quantity%2C%20s_ytd%2C%20s_order_cnt%2C%20s_remote_cnt%2C%20s_data%2C%20s_dist_10%20FROM%20stock%20WHERE%20(s_i_id%2C%20s_w_id)%20IN%20((_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_))%20ORDER%20BY%20s_i_id





---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-:-LFsv0q-4HuWMKMPgN-d:bfjekxl)*

**Comment by knz** [comment]
> can you take a look at this and see if it's getting us closer to what you'd expect here?

I looked at the screenshot and I confirm it is getting us closer to what I'd expect (and what customers would expect).

I will test more once the PR lands.

---

Reviewed 1 of 1 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-:-LFwA1E22XWD66BK5hIU:b-7tgkeu)*

**Comment by vilterp** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ui/src/util/appStats.ts, line 35 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx1xBd4ugEF2MKwae9:-LFx1xBd4ugEF2MKwaeA:b-24gmvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/util/appStats.ts#L35)):*
> ```XML
> }
> 
> export function addStatementStats(a: StatementStatistics, b: StatementStatistics) {
> ```

maybe add a comment that this is duplicated from `app_stats.go`

---

*[pkg/ui/src/views/statements/statementsPage.tsx, line 214 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx2CKl97nZLFnbYGGd:-LFx2CKl97nZLFnbYGGe:bvwx10w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/views/statements/statementsPage.tsx#L214)):*
> ```XML
> }
> 
> const selectStatements = createSelector(
> ```

There's some nontrivial logic in these selectors; would be nice to have a couple of tests. E.g. send in a couple of queries with the same fingerprint but different `failed` or `distsql`; see that they get coalesced correctly.

If not tests demonstrating what this code is doing, please at least add some basic comments above each selector, e.g. "`selectApps` iterates over all fingerprints, returning the set of apps they use" and "`selectStatements` coalesces the returned StatementStats based on fingerprint, not `(fingerprint, distsql, failed)`, as they are originally grouped.". Otherwise, it's pretty hard to understand what's going on here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949)*

**Comment by couchand** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ui/src/util/appStats.ts, line 35 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx1xBd4ugEF2MKwae9:-LG19eMQ6YlWhkyZm0lZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/util/appStats.ts#L35)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

maybe add a comment that this is duplicated from `app_stats.go`
</blockquote></details>

Done.

---

*[pkg/ui/src/views/statements/statementsPage.tsx, line 214 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx2CKl97nZLFnbYGGd:-LG19gc-7xylJDfbmxkW:b-hem8hc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/views/statements/statementsPage.tsx#L214)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

There's some nontrivial logic in these selectors; would be nice to have a couple of tests. E.g. send in a couple of queries with the same fingerprint but different `failed` or `distsql`; see that they get coalesced correctly.

If not tests demonstrating what this code is doing, please at least add some basic comments above each selector, e.g. "`selectApps` iterates over all fingerprints, returning the set of apps they use" and "`selectStatements` coalesces the returned StatementStats based on fingerprint, not `(fingerprint, distsql, failed)`, as they are originally grouped.". Otherwise, it's pretty hard to understand what's going on here.
</blockquote></details>

Good point.  Added short comments and what seem to be reasonably-complete tests.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949)*

**Review by vilterp — APPROVED** [review]
Nice tests & tooltips!

Quick Q: is there supposed to be a little red fleck on the right of the count bar on the table page? 
image

Can't wait to get this on the left nav. I'd say it's about ready, once we have cross-node aggregation!

**Comment by vilterp** [comment]
Suggestion, not blocking this PR: the tooltip for getting the full statement is good, but I still wonder if it would be better to just `overflow: ellipse` that cell. I'd like to be able to e.g. open up the statements page for TPCC and cmd+f to see if there are any joins in it. (Still might miss them if ellipsified, but seems better than the summaries).

**Comment by vilterp** [comment]
I'm skeptical that we'll be able to find some kind of structured summary that's better than the fingerprint itself, ellipsified so that all the rows are the same height. But open to ideas / brainstorming.

**Comment by vilterp** [comment]
Noice! 
image

**Pull request merged**
